src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/ThrowsTaglet.java

```diff
@@ -47,252 +52,733 @@
-                var h = (ExecutableElement) inheritedDoc.holder;
-                inheritedDoc.tagList.forEach(t -> declaredExceptionTags.put((ThrowsTree) t, h));
+    private Map<TypeMirror, TypeMirror> getSubstitutedThrownTypes(Types types,
+                                                                  List<? extends TypeMirror> declaredThrownTypes,
+                                                                  List<? extends TypeMirror> instantiatedThrownTypes) {
+        Map<TypeMirror, TypeMirror> map = new HashMap<>();
+        var i1 = declaredThrownTypes.iterator();
+        var i2 = instantiatedThrownTypes.iterator();
+        while (i1.hasNext() && i2.hasNext()) {
+            TypeMirror t1 = i1.next();
+            TypeMirror t2 = i2.next();
+            if (!types.isSameType(t1, t2)) {
+                map.put(t1, t2);
             }
         }
-        result.add(throwsTagsOutput(declaredExceptionTags, alreadyDocumented, typeSubstitutions,
-                writer));
-        return result;
+        // correspondence between types is established positionally, i.e.
+        // pairwise, which means that the lists must have the same
+        // number of elements; if they don't, this algorithm is
+        // broken
+        assert !i1.hasNext() && !i2.hasNext();
+        // copyOf is unordered and this is fine: this map is for queries;
+        // it doesn't control rendering order
+        return Map.copyOf(map);
     }
 
-    private Content linkToUndocumentedDeclaredExceptions(List<? extends TypeMirror> declaredExceptionTypes,
-                                                         Set<String> alreadyDocumented,
-                                                         TagletWriter writer) {
-        // TODO: assert declaredExceptionTypes are instantiated
-        var utils = writer.configuration().utils;
-        Content result = writer.getOutputInstance();
-        for (TypeMirror declaredExceptionType : declaredExceptionTypes) {
-            TypeElement te = utils.asTypeElement(declaredExceptionType);
-            if (te != null &&
-                    !alreadyDocumented.contains(declaredExceptionType.toString()) &&
-                    !alreadyDocumented.contains(utils.getFullyQualifiedName(te, false))) {
-                if (alreadyDocumented.isEmpty()) {
-                    result.add(writer.getThrowsHeader());
-                }
-                result.add(writer.throwsTagOutput(declaredExceptionType));
-                alreadyDocumented.add(utils.getSimpleName(te));
+    private static AssertionError newAssertionError(Object... objects) {
+        return new AssertionError(Arrays.toString(objects));
+    }
+
+    private static String diagnosticDescriptionOf(Element e) {
+        var name = e instanceof QualifiedNameable q ? q.getQualifiedName() : e.getSimpleName();
+        return name + " (" + detailedDescriptionOf(e) + ")";
+    }
+
+    private static String detailedDescriptionOf(Element e) {
```

Comment:
can't be static if you're going to solve i18n

Reply:
It could wait. I don't have a clear idea of how to do it now. It's also a significant effort. At the same time, if left without diagnostic output, errors could be puzzling.